cohortextractor/patients.py

```diff
@@ -1170,6 +1170,62 @@ def date_of(
     return "value_from", locals()
 
 
+def with_vaccination_record(
+    procedure_codes=None,
+    product_codes=None,
+    # Set date limits
+    on_or_before=None,
+    on_or_after=None,
+    between=None,
+    # Set return type
+    returning="binary_flag",
+    date_format=None,
+    # Matching rule
+    find_first_match_in_period=None,
+    find_last_match_in_period=None,
+    return_expectations=None,
+):
+    """
+    Identify patients with a vaccination record.
+
+    Args:
+        procedure_codes: SNOMED codes indicating the vaccination procedure, eg
+            13906937 (seasonal influenza vaccination).  Note that in the TPP backend we
+            are unable to distinguish between different codes for the same target
+            disease.  For instance, we cannot distinguish between 1324681000000101
+            (first dose of covid vaccine) and 1324691000000104 (second dose of covid
+            vaccine).
```

Comment:
Good points.

Two options:

### 1. Specify target disease instead of the codes used in procedures to vaccinate against the target disease

We could either specify the name of the target disease, or a SNOMED code indicating the target disease.  I think I'd prefer to use a SNOMED code.

Then we'd have a pair of lookups -- in TPP, we'd look up the name of the disease used in the TPP vaccination record, and in EMIS we'd look up the procedures used to vaccinate against the disease.

So you'd have:

```
patients.with_vaccination_record(
    target_disease_code=840539006,  # COVID-19
    product_codes=[39115611000001103],  # Pfizer
)
```

In TPP, we'd use the lookups in this PR to convert those codes to strings; in EMIS we'd convert `target_disease_code` to a list of codes of procedures.

We then lose the ability to query for eg `1324691000000104` (second dose of covid vaccine) distinct from `1324681000000101` (first dose of covid vaccine), but I think this is OK for now, since 0.05% of vaccinated patients have the second dose code and not the first dose code.

It means we won't quite be able to implement the PRIMIS spec faithfully, since `COVADM2_COD` includes the second dose code and while `COVADM1_COD` does not, but this isn't a disaster.

![image](https://user-images.githubusercontent.com/28734/108044907-bf9bb980-703a-11eb-884c-cbf6ecb9d8fa.png)

### 2. Specify arguments for TPP and EMIS backends separately, as in #464 

```
patients.with_vaccination_record(
  tpp={"target_disease_matches": ..., "product_name_matches": ...},
  emis={"procedure_codes": ..., "product_codes": ...},
  ...
)
```

This has the benefit of not requiring any hardcoding, and makes it obvious that the backends behave differently.

Thoughts?

Reply:
I've gone with 2. for now.  This doesn't paint us into any corners, and has been easy to implement.